File: frontend/src/shared/types/game.ts
export type MatchType =
  | 'PLAYER_VS_PLAYER'
  | 'TEAM_VS_TEAM'
  | 'FREE_FOR_ALL'
  | 'SOLO_CHALLENGE'
  | 'COOP_VS_AI';

export type ResultType =
  | 'SCORE'
  | 'TIME'
  | 'PLACEMENT'
  | 'WINNER_ONLY'
  | 'KILLS'
  | 'GOALS'
  | 'ROUNDS_WON'
  | 'CUSTOM_NUMBER';

export type WinnerRule =
  | 'HIGHEST_VALUE_WINS'
  | 'LOWEST_VALUE_WINS'
  | 'FIRST_TO_FINISH_WINS'
  | 'LAST_REMAINING_WINS'
  | 'MOST_ROUNDS_WON'
  | 'MANUAL_WINNER'
  | 'CLOSEST_TO_TARGET';

export type TieBreakerRule =
  | 'ALLOW_DRAW'
  | 'EXTRA_ROUND'
  | 'SUDDEN_DEATH'
  | 'HIGHEST_SECONDARY_VALUE'
  | 'LOWEST_SECONDARY_VALUE'
  | 'MANUAL_DECISION'
  | 'RANDOM';

export type RotationRule =
  | 'NONE'
  | 'CAPTAIN_CHOOSES'
  | 'RANDOM_PLAYER'
  | 'LEAST_PLAYED_PLAYER'
  | 'EVERYONE_MUST_PLAY';

export type BonusCondition =
  | 'PERFECT_SCORE'
  | 'FASTEST_TIME'
  | 'MOST_KILLS'
  | 'BIGGEST_WIN_MARGIN'
  | 'NEW_RECORD'
  | 'CLEAN_SHEET';

export type ScoreRounding = 'NONE' | 'FLOOR' | 'CEIL' | 'ROUND';

export type TimeAction = 'END_MATCH' | 'USE_CURRENT_SCORE' | 'MANUAL_DECISION';

export type ScoringRule =
  | { type: 'WIN_DRAW_LOSS'; pointsForWin: number; pointsForDraw: number; pointsForLoss: number }
  | { type: 'PLACEMENT'; pointsByPlacement: Record<number, number> }
  | { type: 'SCORE_BASED'; multiplier: number; rounding: ScoreRounding }
  | { type: 'WINNER_TAKES_ALL'; pointsToWinner: number }
  | { type: 'MANUAL' };

export type ParticipantRule = {
  minPlayersPerTeam: number;
  maxPlayersPerTeam: number;
  numberOfTeams: number | null;
  allowSubstitutes: boolean;
};

export type ValidationRule = {
  minValue: number | null;
  maxValue: number | null;
  allowDecimals: boolean;
  required: boolean;
};

export type TimeLimitRule = {
  hasTimeLimit: boolean;
  durationMinutes: number | null;
  actionWhenTimeRunsOut: TimeAction;
};

export type BonusRule = {
  name: string;
  condition: BonusCondition;
  bonusPoints: number;
};

export type Game = {
  id: string;
  name: string;
  description: string;
  platform: string | null;
  genre: string | null;
  isActive: boolean;
  matchType: MatchType;
  participantRule: ParticipantRule;
  resultType: ResultType;
  winnerRule: WinnerRule;
  scoringRule: ScoringRule;
  tieBreakerRule: TieBreakerRule;
  validationRule: ValidationRule | null;
  rotationRule: RotationRule | null;
  timeLimitRule: TimeLimitRule | null;
  bonusRules: BonusRule[];
  createdAt: string;
  updatedAt: string;
};

export type GameFormValues = Omit<Game, 'id' | 'createdAt' | 'updatedAt'>;

File: backend/src/main/java/se/backede/domain/model/Game.java
package se.backede.domain.model;

import se.backede.shared.exception.DomainValidationException;

import java.time.Instant;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public record Game(
        UUID id,
        String name,
        String description,
        String platform,
        String genre,
        boolean isActive,
        MatchType matchType,
        ParticipantRule participantRule,
        ResultType resultType,
        WinnerRule winnerRule,
        ScoringRule scoringRule,
        TieBreakerRule tieBreakerRule,
        ValidationRule validationRule,
        RotationRule rotationRule,
        TimeLimitRule timeLimitRule,
        List<BonusRule> bonusRules,
        Instant createdAt,
        Instant updatedAt
) {
    public static final int NAME_MAX_LENGTH = 120;

    public Game {
        Objects.requireNonNull(id, "id must not be null");
        Objects.requireNonNull(matchType, "matchType must not be null");
        Objects.requireNonNull(participantRule, "participantRule must not be null");
        Objects.requireNonNull(resultType, "resultType must not be null");
        Objects.requireNonNull(winnerRule, "winnerRule must not be null");
        Objects.requireNonNull(scoringRule, "scoringRule must not be null");
        Objects.requireNonNull(tieBreakerRule, "tieBreakerRule must not be null");
        Objects.requireNonNull(createdAt, "createdAt must not be null");
        Objects.requireNonNull(updatedAt, "updatedAt must not be null");
        name = name == null ? "" : name.trim();
        description = description == null ? "" : description.trim();
        bonusRules = bonusRules == null ? List.of() : List.copyOf(bonusRules);
        if (name.isBlank()) throw new DomainValidationException("Game name is required");
        if (name.length() > NAME_MAX_LENGTH)
            throw new DomainValidationException("Game name must be at most 120 characters");
    }

    public static Game create(
            String name, String description, String platform, String genre,
            MatchType matchType, ParticipantRule participantRule, ResultType resultType,
            WinnerRule winnerRule, ScoringRule scoringRule, TieBreakerRule tieBreakerRule,
            ValidationRule validationRule, RotationRule rotationRule, TimeLimitRule timeLimitRule,
            List<BonusRule> bonusRules, Instant now) {
        return new Game(UUID.randomUUID(), name, description, platform, genre, true,
                matchType, participantRule, resultType, winnerRule, scoringRule, tieBreakerRule,
                validationRule, rotationRule, timeLimitRule, bonusRules, now, now);
    }

    public static Game rehydrate(
            UUID id, String name, String description, String platform, String genre, boolean isActive,
            MatchType matchType, ParticipantRule participantRule, ResultType resultType,
            WinnerRule winnerRule, ScoringRule scoringRule, TieBreakerRule tieBreakerRule,
            ValidationRule validationRule, RotationRule rotationRule, TimeLimitRule timeLimitRule,
            List<BonusRule> bonusRules, Instant createdAt, Instant updatedAt) {
        return new Game(id, name, description, platform, genre, isActive,
                matchType, participantRule, resultType, winnerRule, scoringRule, tieBreakerRule,
                validationRule, rotationRule, timeLimitRule, bonusRules, createdAt, updatedAt);
    }

    public Game update(
            String name, String description, String platform, String genre, boolean isActive,
            MatchType matchType, ParticipantRule participantRule, ResultType resultType,
            WinnerRule winnerRule, ScoringRule scoringRule, TieBreakerRule tieBreakerRule,
            ValidationRule validationRule, RotationRule rotationRule, TimeLimitRule timeLimitRule,
            List<BonusRule> bonusRules, Instant now) {
        return new Game(id, name, description, platform, genre, isActive,
                matchType, participantRule, resultType, winnerRule, scoringRule, tieBreakerRule,
                validationRule, rotationRule, timeLimitRule, bonusRules, createdAt, now);
    }
}
